File: MTG Deck Editor/src/editor/database/characteristics/Legality.java
package editor.database.characteristics;

/**
 * This enum represents a legality in a format.  If a card is banned, it cannot
 * be played in any deck.  If it is restricted, exactly one copy may appear in a deck
 * and its sideboard.  If it is legal, up to four copies (with the exception of basic
 * lands and cards that say otherwise) may appear in a deck and its sideboard.
 * 
 * @author Alec Roelke
 */
public enum Legality
{
	BANNED("Banned"),
	RESTRICTED("Restricted"),
	LEGAL("Legal");
	
	/**
	 * Parse a String for a Legality.
	 * 
	 * @param s String to parse
	 * @return The Legality corresponding to the contents of the specified String.
	 */
	public static Legality get(String s)
	{
		for (Legality l: Legality.values())
			if (s.equalsIgnoreCase(l.toString()))
				return l;
		throw new IllegalArgumentException("Illegal legality string " + s);
	}
	
	/**
	 * Type of legality a card might have in a format.
	 */
	private final String legality;
	
	/**
	 * Create a new Legality.
	 * 
	 * @param legality Type of legality a card might have.
	 */
	private Legality(final String legality)
	{
		this.legality = legality;
	}
	
	/**
	 * @return A String representation of this Legality.
	 */
	@Override
	public String toString()
	{
		return legality;
	}
}

File: src/main/scala/editor/database/attributes/Legality.scala
package editor.database.attributes

object Legality {
  def parse(s: String) = values.find(_.toString.equalsIgnoreCase(s)).headOption
}

/**
 * Legality of a card in a format. A legal card can be restricted to having only one copy in a deck, which
 * is mainly used for Vintage (singleton-only formats like Commander don't count). Illegal cards are broken
 * down into two groups: cards that are illegal (i.e. the expansion or printing is not legal) and cards that are
 * banned (they would be legal except they have been declared illegal for various reasons).
 * 
 * @constructor create a new legality type
 * @param toString string representation of the legality
 * @param isLegal whether or not it represents a legal card in a format
 * 
 * @author Alec Roelke
 */
enum Legality(override val toString: String, val isLegal: Boolean) {
  /** A card that has been banned from a format. */
  case BANNED extends Legality("Banned", false)
  /** A card that is legal in a format. */
  case LEGAL extends Legality("Legal", true)
  /** A card that is legal, but restricted in a format. */
  case RESTRICTED extends Legality("Restricted", true)
  /** A card that is illegal in a format. */
  case ILLEGAL extends Legality("Illegal", false)
}
